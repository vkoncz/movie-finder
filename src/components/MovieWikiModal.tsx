import {
  IonButton,
  IonCol,
  IonContent,
  IonIcon,
  IonRow,
  IonSkeletonText,
  IonTitle,
  IonToolbar,
} from '@ionic/react';
import React, { useEffect, useState } from 'react';
import { wikiSummaryClient } from '../data/fetch/wikiSummaryClient';
import { chevronBack, openOutline } from 'ionicons/icons';
import { MovieInfo } from '../models/MovieInfo';

interface Props {
  movieInfo: MovieInfo;
  modalClose: () => void;
}

export const MovieWikiModal: React.FC<Props> = ({ movieInfo, modalClose }) => {
  const [wikiSummary, setWikiSummary] = useState('');
  const [wikiLink, setWikiLink] = useState('');
  const [isLoading, setIsLoading] = useState(false);
  const [missingWiki, setMissingWiki] = useState(false);

  useEffect(() => {
    setIsLoading(true);
    setMissingWiki(false);

    wikiSummaryClient(movieInfo.title)
      .then(result => {
        setWikiSummary(result.extract);
        setWikiLink(result.content_urls.desktop.page);
        setIsLoading(false);
      })
      .catch(() => {
        setIsLoading(false);
        setMissingWiki(true);
      });
  }, [movieInfo]);

  return (
    <>
      <IonToolbar color="medium">
        <IonTitle>{movieInfo.title}</IonTitle>
      </IonToolbar>
      <IonContent>
        <p className="ion-padding-start ion-padding-end">
          {isLoading ? (
            <>
              <IonSkeletonText animated style={{ width: '60%' }} />
              <IonSkeletonText animated />
              <IonSkeletonText animated style={{ width: '90%' }} />
              <IonSkeletonText animated style={{ width: '70%' }} />
            </>
          ) : (
            wikiSummary
          )}
          {missingWiki ? 'Cannot found Wikipedia article' : ''}
        </p>

        {!isLoading && !missingWiki ? (
          <IonRow>
            <IonButton href={wikiLink} target="_blank" color="light">
              Open Wikipedia article
              <IonIcon icon={openOutline} />
            </IonButton>
          </IonRow>
        ) : (
          ''
        )}

        <IonRow>
          <IonButton
            href={`https://www.imdb.com/title/${movieInfo.imdb}`}
            target="_blank"
            color="light"
          >
            IMDB
            <IonIcon icon={openOutline} />
          </IonButton>
        </IonRow>

        <IonRow>
          <IonCol class="ion-text-center">
            <IonButton
              color="warning"
              class="ion-align-self-center"
              shape="round"
              onClick={modalClose}
            >
              <IonIcon icon={chevronBack} />
            </IonButton>
          </IonCol>
        </IonRow>
      </IonContent>
    </>
  );
};
