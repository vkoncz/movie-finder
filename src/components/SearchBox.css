ion-button {
  height: 42px;
  width: 60px;
}

ion-grid {
  max-width: 1000px;
}
