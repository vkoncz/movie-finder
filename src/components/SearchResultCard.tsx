import {
  IonCard,
  IonCardContent,
  IonCardHeader,
  IonCardTitle,
  IonIcon,
  IonItem,
  IonLabel,
} from '@ionic/react';
import React from 'react';
import { Details } from '../models/MovieDetails';
import { star, pricetag } from 'ionicons/icons';

interface Props {
  details: Details;
  onMovieClick: (title: string) => void;
}

export const SearchResultCard: React.FC<Props> = ({ details, onMovieClick }) => {
  return (
    <IonCard
      button={true}
      onClick={() => {
        onMovieClick(details.title);
      }}
    >
      <IonCardHeader color="medium">
        <IonCardTitle>{details.title}</IonCardTitle>
      </IonCardHeader>

      <IonCardContent>
        <IonItem lines="none">
          <IonIcon icon={star} slot="start" size="small"></IonIcon>
          <IonLabel>{details.rating}</IonLabel>
        </IonItem>
        <IonItem lines="none">
          <IonIcon icon={pricetag} slot="start" size="small"></IonIcon>
          <IonLabel>{details.genres.map(genre => genre.name).join(', ')}</IonLabel>
        </IonItem>
      </IonCardContent>
    </IonCard>
  );
};
