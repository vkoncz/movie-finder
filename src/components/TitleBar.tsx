import React from 'react';
import { IonTitle, IonToolbar } from '@ionic/react';

export const TitleBar: React.FC = () => {
  return (
    <IonToolbar color="primary">
      <IonTitle>Movie Finder</IonTitle>
    </IonToolbar>
  );
};
