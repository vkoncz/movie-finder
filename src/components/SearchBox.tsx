import React, { useState } from 'react';
import {
  IonButton,
  IonCol,
  IonRow,
  IonGrid,
  IonIcon,
  IonSearchbar,
  IonToast,
  IonLoading,
} from '@ionic/react';
import { search, videocam } from 'ionicons/icons';
import './SearchBox.css';
import { Details } from '../models/MovieDetails';
import { graphSearchClient } from '../data/fetch/graphSearchClient';

interface Props {
  resultChange: (results?: Details[]) => void;
}

export const SearchBox: React.FC<Props> = ({ resultChange }) => {
  const [searchTerm, setSearchTerm] = useState('');
  const [showError, setShowError] = useState(false);
  const [isLoading, setIsLoading] = useState(false);

  const fetchData = () => {
    setIsLoading(true);
    graphSearchClient(searchTerm)
      .then(result => {
        resultChange(result);
        setIsLoading(false);
      })
      .catch(error => {
        console.log('Connection error:', error);
        setShowError(true);
        setIsLoading(false);
      });
  };

  return (
    <>
      <IonGrid class="ion-center search-container">
        <IonRow class="ion-align-items-center">
          <IonCol class="ion-no-padding">
            <IonSearchbar
              autocomplete="on"
              placeholder="Search movies"
              searchIcon={videocam}
              value={searchTerm}
              onIonChange={e => setSearchTerm(e.detail.value!)}
              onKeyUp={e => {
                if (e.key === 'Enter') fetchData();
              }}
            />
          </IonCol>
          <IonCol size="auto" class="ion-no-padding ion-align-self-center">
            <IonButton onClick={fetchData} class="search-button">
              <IonIcon icon={search} />
            </IonButton>
          </IonCol>
        </IonRow>
      </IonGrid>

      <IonLoading isOpen={isLoading} message="Searching movies..." />

      <IonToast
        isOpen={showError}
        onDidDismiss={() => setShowError(false)}
        message="Connection error, please try it again later"
        duration={5000}
      />
    </>
  );
};
