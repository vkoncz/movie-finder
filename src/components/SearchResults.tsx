import { IonCol, IonGrid, IonModal, IonRow, IonText } from '@ionic/react';
import React, { useState } from 'react';
import { Details } from '../models/MovieDetails';
import { MovieInfo } from '../models/MovieInfo';
import { MovieWikiModal } from './MovieWikiModal';
import { SearchResultCard } from './SearchResultCard';

interface Props {
  results?: Details[];
}

export const SearchResults: React.FC<Props> = ({ results }) => {
  const message = !results || results?.length ? '' : 'No results found';
  const [selectedMovieInfo, setSelectedMovieInfo] = useState<MovieInfo>({ imdb: '', title: '' });
  const [modalVisible, setModalVisible] = useState(false);

  return (
    <>
      <IonGrid>
        <IonRow>
          {results
            ? results.map((result, key) => (
                <IonCol key={key} sizeXl="3" sizeLg="4" sizeMd="6" size="12">
                  <SearchResultCard
                    details={result}
                    onMovieClick={title => {
                      setModalVisible(true);
                      setSelectedMovieInfo(title);
                    }}
                  />
                </IonCol>
              ))
            : ''}
          <IonText>{message}</IonText>
        </IonRow>
      </IonGrid>

      <IonModal isOpen={modalVisible} onDidDismiss={() => setModalVisible(false)}>
        <MovieWikiModal movieInfo={selectedMovieInfo} modalClose={() => setModalVisible(false)} />
      </IonModal>
    </>
  );
};
