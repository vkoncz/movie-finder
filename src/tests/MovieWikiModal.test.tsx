import React from 'react';
import { render } from '@testing-library/react';
import { MovieWikiModal } from '../components/MovieWikiModal';
import { movieInfo } from './test-data/movieInfo';

// TODO: mock wikiSummaryClient with test-data/wikiSummary.ts
it('should match to MovieWikiModal snapshots', () => {
  const { asFragment } = render(<MovieWikiModal movieInfo={movieInfo} modalClose={() => {}} />);

  expect(asFragment()).toMatchSnapshot();
});
