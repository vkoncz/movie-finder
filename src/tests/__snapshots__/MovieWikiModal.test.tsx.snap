// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match to MovieWikiModal snapshots 1`] = `
<DocumentFragment>
  <ion-toolbar
    color="medium"
  >
    <ion-title>
      The Lord of the Rings: The Return of the King
    </ion-title>
  </ion-toolbar>
  <ion-content>
    <p
      class="ion-padding-start ion-padding-end"
    >
      <ion-skeleton-text
        style="width: 60%;"
      />
      <ion-skeleton-text />
      <ion-skeleton-text
        style="width: 90%;"
      />
      <ion-skeleton-text
        style="width: 70%;"
      />
    </p>
    <ion-row>
      <ion-button
        color="light"
        href="https://www.imdb.com/title/tt0167260"
        target="_blank"
      >
        IMDB
        <ion-icon
          icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Open</title><path d='M384 224v184a40 40 0 01-40 40H104a40 40 0 01-40-40V168a40 40 0 0140-40h167.48M336 64h112v112M224 288L440 72' stroke-linecap='round' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
        />
      </ion-button>
    </ion-row>
    <ion-row>
      <ion-col
        class="ion-text-center"
      >
        <ion-button
          class="ion-align-self-center"
          color="warning"
          shape="round"
        >
          <ion-icon
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Chevron Back</title><path stroke-linecap='round' stroke-linejoin='round' stroke-width='48' d='M328 112L184 256l144 144' class='ionicon-fill-none'/></svg>"
          />
        </ion-button>
      </ion-col>
    </ion-row>
  </ion-content>
</DocumentFragment>
`;
